source/MessageLogPanel.cpp

```diff
@@ -0,0 +1,147 @@
+/* MessageLogPanel.cpp
+Copyright (c) 2023 by TomGoodIdea
+
+Endless Sky is free software: you can redistribute it and/or modify it under the
+terms of the GNU General Public License as published by the Free Software
+Foundation, either version 3 of the License, or (at your option) any later version.
+
+Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE. See the GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License along with
+this program. If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#include "MessageLogPanel.h"
+
+#include "text/alignment.hpp"
+#include "Color.h"
+#include "Command.h"
+#include "FillShader.h"
+#include "text/Font.h"
+#include "text/FontSet.h"
+#include "GameData.h"
+#include "Preferences.h"
+#include "Screen.h"
+#include "Sprite.h"
+#include "SpriteSet.h"
+#include "SpriteShader.h"
+#include "UI.h"
+#include "text/WrappedText.h"
+
+using namespace std;
+
+namespace {
+	const double PAD = 10.;
+	const double LINE_HEIGHT = 25.;
+}
+
+
+
+MessageLogPanel::MessageLogPanel(const vector<pair<string, Messages::Importance>> &messages)
+	: messages(messages)
+{
+	SetInterruptible(false);
+}
+
+
+
+// Draw this panel.
+void MessageLogPanel::Draw()
+{
+	// Dim out everything outside this panel.
+	DrawBackdrop();
+
+	// Draw the panel.
+	width = Screen::Width() - 300;
+	const Color &backColor = *GameData::Colors().Get("message log background");
+	FillShader::Fill(
+		Point(Screen::Left() + .5 * width, 0.),
+		Point(width, Screen::Height()),
+		backColor);
+
+	const Sprite *edgeSprite = SpriteSet::Get("ui/right edge");
+	if(edgeSprite->Height())
+	{
+		// If the screen is high enough, the edge sprite should repeat.
+		double spriteHeight = edgeSprite->Height();
+		Point pos(
+			Screen::Left() + width + .5 * edgeSprite->Width(),
+			Screen::Top() + .5 * spriteHeight);
+		for( ; pos.Y() - .5 * spriteHeight < Screen::Bottom(); pos.Y() += spriteHeight)
+			SpriteShader::Draw(edgeSprite, pos);
+	}
+
+	const Font &font = FontSet::Get(14);
+
+	// Parameters for drawing messages:
+	WrappedText messageLine(font);
+	messageLine.SetAlignment(Alignment::LEFT);
+	messageLine.SetWrapWidth(width - 2. * PAD);
+
+	// Draw messages.
+	Point pos = Screen::BottomLeft() + Point(PAD, scroll);
+	for(const auto &it : messages)
+	{
+		messageLine.Wrap(it.first);
+		pos.Y() -= messageLine.Height();
+		messageLine.Draw(pos, *Messages::GetColor(it.second, true));
+	}
```

Comment:
arent you also drawing messages that cannot be seen?
just add a check if the y pos is still in bounds, for 10000 messages that would be quite the amount of draw calls